README.md
# WinDjView
*A fast, compact, and powerful DjVu viewer.*

Brought to you by: [Andrew Zhezherun](https://sourceforge.net/u/zhezherun/profile/)

## Summary
WinDjView is a fast, compact and powerful DjVu viewer for Windows with tabbed interface, continuous scrolling and advanced printing options, based on free DjVuLibre library.

![WinDjView screenshot](https://a.fsdn.com/con/app/proj/windjview/screenshots/328149.jpg)

## Files
[Download WinDjView](https://sourceforge.net/projects/windjview/files/latest/download)

## Support
[Discussion](https://sourceforge.net/p/windjview/discussion/)

## License
GNU General Public License version 2.0 (GPLv2)

## Code
This repository was kindly converted from an official [SourceForge.Net CVS-repository](https://sourceforge.net/p/windjview/code/) with a preserved commit history.


windjview/changes.txt
WinDjView version history
Copyright (C) 2004-2015 Andrew Zhezherun
----------------------------------------

Changes in version 2.1 (2015-02-23):
+ Restore last open tabs on startup.
* Rebuilt with VS2013. Windows XP or later is now required.
* Some minor bug fixes.

Changes in version 2.0.2 (2012-10-21):
* Fixed an infinite loop in libdjvu when parsing some annotations.
* Fixed an issue when subpixel scaling was requested but not used.

Changes in version 2.0.1 (2012-07-01):
! New toolbar icons and interface tweaks (contributed by
  Leonid Zhezherun).
! Native 64-bit executable for Windows 7/Vista 64-bit.
  Windows 2000 or later is now required. Windows 98/ME users
  should continue to use version 1.0.3.
+ Optional subpixel page scaling for better looking results
  on some LCD monitors (contributed by Larry Davis).
+ Smoother mouse wheel panning and autoscrolling.
* Fixed a crash when adding bookmarks in fullscreen mode.
* Several minor fixes.

Changes in version 1.0.3 (2009-08-15):
* Fixed a problem with exiting fullscreen mode introduced in 1.0.2.

Changes in version 1.0.2 (2009-08-14):
+ Export multiple pages at once via multiple selection in the
  thumbnails view.
+ Support for additional image export formats: PNG, GIF, TIF, JPG
  (Windows XP or later required, or GDI+ on older versions).
+ Each document now has its own history (back/forward buttons).
* Fixed touchpad scrolling on some laptops.
* Several bug fixes.

Changes in version 1.0.1 (2009-03-20):
+ Tabs can be closed with wheel click or middle click.
+ Use new style file dialogs on Windows 7/Vista.
+ File open and save dialogs can be resized.
* Fixed a problem with bookmarks not working immediately after import.
* Several minor fixes.

Changes in version 1.0 (2009-02-19):
! Finally, version number starts with one! This means that the program
  is officially considered stable after years of development.
! Added multiple top-level documents mode.
! Added a tab bar to the single-window mode.
+ Added an option to enable continuous scrolling in fullscreen mode.
+ Added an option to order pages from right to left in facing mode.
+ User-created bookmarks and annotations can be imported and exported.
+ More embedded hyperlink and annotation types are supported.
+ Drag mode is activated when right mouse button is pressed.
+ Added support for dictionaries that look up the word under mouse pointer.
+ Thumbnail size can now be enlarged or reduced.
+ Document properties can be accessed from the File menu.
+ Added F3 hotkey for Find Next.
+ Added support for Unicode Supplementary Planes in contents and index.
+ DPI value is now set when exporting page or selection as bitmap.
+ Added a warning if WinDjView is not the default DjVu viewer.
+ Added an automatic weekly check for updates.
* Fixed saving "First page alone" setting for every document.
* Fixed crash in magnify tool after rotating the page.
* Fixed crash when exiting fullscreen mode.
* Fixed crash when compiled with Visual Studio 2005.
* Many other small fixes and improvements.

Changes in version 0.5 (2007-05-30):
! Implemented user-created bookmarks and annotations.
! Added rectangle tool: can be used to create an annotation,
  to zoom in, to print selected region or to export it to a bitmap file.
! Integrated dictionary features of DjVuDic program.
+ Open page and view settings can be restored when opening a file.
+ Page number can be specified when opening files via command line.
+ Back/Forward keys on multimedia keyboards now work as Alt+arrows.
+ Reorganized Settings dialog.
+ Added an option to invert colors.
+ Added an option to change display units (centimeters, millimeters, inches).
+ New printing options: auto-rotate pages, booklet printing.
* Improved interface responsiveness when opening and closing documents.
* Improvements for files with large shared dictionaries:
  * Reduced memory usage.
  * Fixed interface stalls.
  * Increased decoding speed.
* Fixed slow loading of large bookmark trees.
* Fixed printing in two pages per sheet mode.
* Fixed a crash in print dialog with some printers.
* A number of other small fixes and improvements.

Changes in version 0.4.3 (2006-09-15):
+ Added panning when mouse wheel is pressed.
* Improved performance of magnifying glass.
* Fixed displaying UTF-8 bookmarks on pre-XP systems.
* Fixed problem with long documents on Win98 when thumbnails are on.
* Fixed font issues.
* Several crash fixes.

Changes in version 0.4.2 (2006-07-24):
! Added magnifying glass tool.
+ Added an option to wrap long bookmarks.
+ Added an option to hide scrollbars in fullscreen mode.
+ Added an option to invert direction of Ctrl+Mouse Wheel zooming.
+ Added an option to close the program when Escape is pressed.
+ Added keyboard shortcuts to rotate pages, change layouts and open settings.
* Fixed Open File dialog (.djv files were not displayed).
* Fixed print dialog: collate option is now always available.
* Fixed printing order for custom page ranges.
* Fixed a problem with XP themes.
* Several crash fixes, including a patch from KViewShell DjVu plugin.

Changes in version 0.4.1 (2006-02-25):
! Added support for JPEG-encoded DjVu files.
+ Added an option to select different scaling algorithm
  (slower, but gives better results on low zoom levels).
+ Reduced redrawing when switching documents.
+ Added autoscroll when selecting text by dragging.
+ Rendered pages are reused when entering/exiting fullscreen mode.
+ Added Home/End keyboard shortcuts.
+ Added an option to warn when closing multiple documents.
* Fixed saving and restoring maximized main window state.
* Fixed displaying floating-point numbers in non-English locales.
* Fixed text selection when a page has nonzero original rotation.
* Fixed page caching (caused high CPU and memory usage in some situations).
* Fixed displaying tooltips for annotations in fullscreen mode.

Changes in version 0.4 (2006-01-18):
! Unicode build for Windows 2k/XP.
! Added support for switching user interface languages.

windjview/Global.h
//	WinDjView
//	Copyright (C) 2004-2015 Andrew Zhezherun
//
//	This program is free software; you can redistribute it and/or modify
//	it under the terms of the GNU General Public License as published by
//	the Free Software Foundation; either version 2 of the License, or
//	(at your option) any later version.
//
//	This program is distributed in the hope that it will be useful,
//	but WITHOUT ANY WARRANTY; without even the implied warranty of
//	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//	GNU General Public License for more details.
//
//	You should have received a copy of the GNU General Public License along
//	with this program; if not, write to the Free Software Foundation, Inc.,
//	51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
//	http://www.gnu.org/copyleft/gpl.html

#pragma once


class RefCount
{
public:
	RefCount() : m_nRefCount(1) {}
	virtual ~RefCount() = 0;

	virtual void AddRef()
	{
		InterlockedIncrement(&m_nRefCount);
	}

	virtual void Release()
	{
		if (InterlockedDecrement(&m_nRefCount) <= 0)
			delete this;
	}

protected:
	long m_nRefCount;
};


struct Message
{
	Message(int c) : code(c) {}
	
	int code;
};

class Observable;

class Observer
{
public:
	virtual ~Observer() {}
	virtual void OnUpdate(const Observable* source, const Message* message) = 0;
};

class Observable
{
public:
	void AddObserver(Observer* observer);
	void RemoveObserver(Observer* observer);
	bool HasObservers() const { return !m_observers.empty(); }
	bool IsObservedBy(Observer* observer) const
		{ return m_observers.find(observer) != m_observers.end(); }
	void UpdateObservers(const Message& message);
	
protected:
	set<Observer*> m_observers;
};

bool IsWinVistaOrLater();

void MakeWString(const CString& strText, wstring& result);
bool MakeWString(const GUTF8String& text, wstring& result);
void MakeANSIString(const wstring& text, string& result);
void MakeANSIString(const CString& strText, string& result);
CString MakeCString(const wstring& text);
CString MakeCString(const GUTF8String& text);
GUTF8String MakeUTF8String(const CString& strText);
GUTF8String MakeUTF8String(const wstring& strText);

bool MoveToTrash(LPCTSTR lpszFileName);

void CreateSystemDialogFont(CFont& font);
void CreateSystemIconFont(CFont& font);
void CreateSystemMenuFont(CFont& font);

UINT GetMouseScrollLines();

CRect GetMonitorWorkArea(const CPoint& point);
CRect GetMonitorWorkArea(CWnd* pWnd);
CRect GetMonitorRect(CWnd* pWnd);

bool IsFromCurrentProcess(CWnd* pWnd);

int CompareVersions(const CString& strFirst, const CString& strSecond);
CString FormatDouble(double fValue);
void AFXAPI DDX_MyText(CDataExchange* pDX, int nIDC, double& value, double def = 0.0, LPCTSTR pszSuffix = NULL);
void AFXAPI DDX_MyText(CDataExchange* pDX, int nIDC, DWORD& value, DWORD def = 0, LPCTSTR pszSuffix = NULL);

void SendMessageToVisibleDescendants(HWND hWnd, UINT message, WPARAM wParam = 0, LPARAM lParam = 0);

inline CRect GetClientRect(const CWnd* pWnd)
{
	ASSERT(pWnd != NULL);
	CRect rect;
	pWnd->GetClientRect(rect);
	return rect;
}

inline CRect GetClientRect(const CWnd& wnd)
	{ return GetClientRect(&wnd); }

inline CSize GetClientSize(const CWnd* pWnd)
{
	ASSERT(pWnd != NULL);
	CRect rect;
	pWnd->GetClientRect(rect);
	return rect.Size();
}

inline CSize GetClientSize(const CWnd& wnd)
	{ return GetClientSize(&wnd); }

#define PAGE_RENDERED 1
#define PAGE_DECODED 2
#define LINK_CLICKED 3
#define SEARCH_RESULT_CLICKED 4
#define THUMBNAIL_RENDERED 5
#define THUMBNAIL_CLICKED 6
#define CURRENT_PAGE_CHANGED 7
#define ROTATE_CHANGED 8
#define VIEW_ACTIVATED 9
#define ZOOM_CHANGED 10
#define APP_SETTINGS_CHANGED 11
#define APP_LANGUAGE_CHANGED 12
#define BOOKMARK_ADDED 13
#define ANNOTATION_ADDED 14
#define BOOKMARK_DELETED 15
#define ANNOTATION_DELETED 16
#define BOOKMARK_CLICKED 17
#define VIEW_INITIALIZED 18
#define SOURCE_RELEASED 19
#define KEY_STATE_CHANGED 20
#define DICT_LIST_CHANGED 21
#define SIDEBAR_TAB_CHANGED 22
#define TAB_ACTIVATING 23
#define TAB_ACTIVATED 24
#define TAB_CLOSED 25
#define ANNOTATIONS_CHANGED 26
#define BOOKMARKS_CHANGED 27
#define PRINT_PAGES 28
#define EXPORT_PAGES 29

class CDIB;
struct Bookmark;
struct Annotation;

struct PageMsg : public Message
{
	PageMsg(int msg, int nPage_)
		: Message(msg), nPage(nPage_) {}

	int nPage;
};

struct BitmapMsg : public Message
{
	BitmapMsg(int msg, int nPage_, CDIB* pDIB_)
		: Message(msg), nPage(nPage_), pDIB(pDIB_) {}

	int nPage;
	CDIB* pDIB;
};

struct LinkClicked : public Message
{
	LinkClicked(const GUTF8String& url_)
		: Message(LINK_CLICKED), url(url_) {}

	const GUTF8String& url;
};

struct SearchResultClicked : public Message
{
	SearchResultClicked(int nPage_, int nSelStart_, int nSelEnd_)
		: Message(SEARCH_RESULT_CLICKED), nPage(nPage_), nSelStart(nSelStart_), nSelEnd(nSelEnd_) {}

	int nPage, nSelStart, nSelEnd;
};

struct RotateChanged : public Message
{
	RotateChanged(int nRotate_)
		: Message(ROTATE_CHANGED), nRotate(nRotate_) {}

	int nRotate;
};

struct AnnotationMsg : public Message
{
	AnnotationMsg(int msg, Annotation* pAnno_, int nPage_)
		: Message(msg), pAnno(pAnno_), nPage(nPage_) {}

	Annotation* pAnno;
	int nPage;
};

struct BookmarkMsg : public Message
{
	BookmarkMsg(int msg, const Bookmark* pBookmark_)
		: Message(msg), pBookmark(pBookmark_) {}

	const Bookmark* pBookmark;
};

struct KeyStateChanged : public Message
{
	KeyStateChanged(UINT nKey_, bool bPressed_)
		: Message(KEY_STATE_CHANGED), nKey(nKey_), bPressed(bPressed_) {}

	UINT nKey;
	bool bPressed;
};

struct TabMsg : public Message
{
	TabMsg(int msg, CWnd* pWnd_, int nTab_)
		: Message(msg), pWnd(pWnd_), nTab(nTab_) {}

	CWnd* pWnd;
	int nTab;
};

struct PageRangeMsg : public Message
{
	PageRangeMsg(int msg, const set<int>& pages_)
		: Message(msg), pages(pages_) {}

	const set<int>& pages;
};


struct MD5
{
	MD5();
	MD5(const void* data, size_t len);
	MD5(const MD5& md5);

	void Append(const void* data, size_t len);
	void Finish();

	CString ToString() const;
	bool operator==(const MD5& rhs) const
		{ return memcmp(md, rhs.md, sizeof(md)) == 0; }
	bool operator<(const MD5& rhs) const
		{ return memcmp(md, rhs.md, sizeof(md)) < 0; }
	
private:
	void Block(const void* data);

	struct State;
	State* state;

	unsigned char md[16];
};


string& Base64Encode(string& s);
string& Base64Decode(string& s);

inline CString LoadString(UINT nID)
{
	CString strResult;
	strResult.LoadString(nID);
	return strResult;
}

inline CString FormatString(LPCTSTR pszFormat, ...)
{
	CString strResult;

	va_list argList;
	va_start(argList, pszFormat);
	strResult.FormatV(pszFormat, argList);
	va_end(argList);

	return strResult;
}

inline CString FormatString(UINT nFormatID, ...)
{
	CString strResult, strFormat;
	VERIFY(strFormat.LoadString(nFormatID));

	va_list argList;
	va_start(argList, nFormatID);
	strResult.FormatV(strFormat, argList);
	va_end(argList);

	return strResult;
}

inline int GetTotalRotate(GP<DjVuImage> pImage, int nRotate)
{
	GP<DjVuInfo> info = pImage->get_info();
	return (nRotate + (info != NULL ? info->orientation : 0)) % 4;
}
